File: Data Structures & Algorithms/anagram-groups/submission-0.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_map = dict()
        for s in strs:
            tuple_val = self.getCount(s)
            if tuple_val not in anagram_map:
                anagram_map[tuple_val] = []
            anagram_map[tuple_val].append(s)
        return list(anagram_map.values())


    def getCount(self, s: str) -> tuple:
        count = [0]*26
        for c in s:
            count[ord(c) - ord('a')] +=1
        return tuple(count)

File: Data Structures & Algorithms/anagram-groups/submission-0.go
func groupAnagrams(strs []string) [][]string {
	anagramMap := make(map[[26]int][]string)
	for _,s := range strs {
		hashVal := getHash(s)
		anagramMap[hashVal] = append(anagramMap[hashVal],s)
	}
	result := make([][]string, 0, len(anagramMap))
	for _, group := range anagramMap {
		result = append(result, group)
	} 

	return result
}

func getHash(str string) [26]int {
	str = strings.ToLower(str)
	var arr [26]int // properly initializes all 26 slots to 0
	for _, ch := range str {
		// Ensure we only count valid lowercase letters to prevent panics/out-of-bounds
		if ch >= 'a' && ch <= 'z' {
			arr[ch-'a']++
		}
	}
	return arr
}
